pkg/syncer/namespace/namespace_controller.go

```diff
@@ -0,0 +1,164 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+)
+
+const (
+	controllerName = "kcp-workload-syncer-namespace"
+)
+
+type Controller struct {
+	queue workqueue.RateLimitingInterface
+
+	upstreamClient      dynamic.ClusterInterface
+	downstreamClient    dynamic.Interface
+	upstreamInformers   dynamicinformer.DynamicSharedInformerFactory
+	downstreamInformers dynamicinformer.DynamicSharedInformerFactory
+
+	syncTargetName      string
+	syncTargetWorkspace logicalcluster.Name
+	syncTargetUID       types.UID
+	syncTargetKey       string
+}
+
+func NewNamespaceSyncer(syncTargetWorkspace logicalcluster.Name, syncTargetName, syncTargetKey string, upstreamClient dynamic.ClusterInterface, downstreamClient dynamic.Interface,
+	upstreamInformers, downstreamInformers dynamicinformer.DynamicSharedInformerFactory, syncTargetUID types.UID) (*Controller, error) {
+
+	c := Controller{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+
+		upstreamClient:      upstreamClient,
+		downstreamClient:    downstreamClient,
+		upstreamInformers:   upstreamInformers,
+		downstreamInformers: downstreamInformers,
+
+		syncTargetName:      syncTargetName,
+		syncTargetWorkspace: syncTargetWorkspace,
+		syncTargetUID:       syncTargetUID,
+		syncTargetKey:       syncTargetKey,
+	}
+
+	namespaceGVR := schema.GroupVersionResource{
+		Group:    "",
+		Version:  "v1",
+		Resource: "namespaces",
+	}
+
+	// React when there's a namespace deletion upstream.
+	c.upstreamInformers.ForResource(namespaceGVR).Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			c.AddToQueue(namespaceGVR, obj)
+		},
+	})
+
+	// Those handlers are for start/resync cases, in case a namespace deletion event is missed, these handlers
+	// will make sure that we cleanup the namespace in downstream after restart/resync.
+	c.downstreamInformers.ForResource(namespaceGVR).Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.AddToQueue(namespaceGVR, obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.AddToQueue(namespaceGVR, newObj)
+		},
+	})
+
+	klog.V(2).InfoS("Set up upstream namespace informer", "syncTargetWorkspace", syncTargetWorkspace, "syncTargetName", syncTargetName, "syncTargetKey", syncTargetKey)
```

Comment:
Please use the new contextual logging patterns found in other controllers that have already been converted (look for `logger`)

Reply:
updated, PTAL